sync/task_pool.go

```diff
@@ -158,15 +177,53 @@ func (p *TaskPool) run(id int, q chan task) error {
 	}
 }
 
-// add task
-func (p *TaskPool) AddTask(t task) {
-	id := atomic.AddUint32(&p.idx, 1) % uint32(p.tQNumber)
+// AddTask wait idle worker add task
+// return false when the pool is stop
+func (p *TaskPool) AddTask(t task) (ok bool) {
+	idx := atomic.AddUint32(&p.idx, 1)
+	id := idx % uint32(p.tQNumber)
 
 	select {
 	case <-p.done:
-		return
+		return false
+	default:
+		p.qArray[id] <- t
+		return true
+	}
+}
+
+// AddTaskAlways add task to queues or do it immediately
+func (p *TaskPool) AddTaskAlways(t task) {
+	id := atomic.AddUint32(&p.idx, 1) % uint32(p.tQNumber)
+
+	select {
```

Comment:
is it need the check p.done?

Reply:
the method `xxAlways` means it will called anyway